plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -2905,12 +2911,46 @@ public boolean delegateMaterializedViewRefreshToConnector(ConnectorSession sessi
     }
 
     @Override
-    public ConnectorInsertTableHandle beginRefreshMaterializedView(ConnectorSession session, ConnectorTableHandle tableHandle, List<ConnectorTableHandle> sourceTableHandles, RetryMode retryMode)
+    public ConnectorInsertTableHandle beginRefreshMaterializedView(ConnectorSession session, ConnectorTableHandle tableHandle, List<ConnectorTableHandle> sourceTableHandles, RetryMode retryMode, RefreshType refreshType)
     {
+        checkState(fromSnapshotsForRefresh.isEmpty(), "Snapshot map must be empty at the start of MV refresh operation.");
         IcebergTableHandle table = (IcebergTableHandle) tableHandle;
         Table icebergTable = catalog.loadTable(session, table.getSchemaTableName());
         beginTransaction(icebergTable);
 
+        fromSnapshotsForRefresh = Optional.of(new HashMap<>());
+        Optional<String> dependencies = Optional.ofNullable(icebergTable.currentSnapshot())
+                .map(Snapshot::summary)
+                .map(summary -> summary.get(DEPENDS_ON_TABLES));
+
+        boolean shouldUseIncremental = isIncrementalRefreshEnabled(session)
+                && refreshType == RefreshType.INCREMENTAL
+                // there is a single source table (we compare to 2 instead of 1 because the MV storage table is also in the list)
+                && sourceTableHandles.size() == 2
+                // and source table is an Iceberg table
+                && sourceTableHandles.stream().allMatch(tbl -> tbl instanceof IcebergTableHandle)
+                // and the source table's fromSnapshot is available in the MV snapshot summary
+                && dependencies.isPresent() && !dependencies.get().equals(UNKNOWN_SNAPSHOT_TOKEN);
+
+        if (shouldUseIncremental) {
+            Map<String, String> baseTableToSnapshot = MAP_SPLITTER.split(dependencies.get());
+            ImmutableMap.Builder<SchemaTableName, Long> fromSnapshotsForRefreshBuilder = ImmutableMap.builder();
+            for (Map.Entry<String, String> baseTable : baseTableToSnapshot.entrySet()) {
+                String[] schemaTable = baseTable.getKey().split("\\.");
+                SchemaTableName sourceSchemaTable = new SchemaTableName(schemaTable[0], schemaTable[1]);
+                // we have already verified above that there is a single, Iceberg source table
+                IcebergTableHandle sourceTableHandle = (IcebergTableHandle) sourceTableHandles.getFirst();
+                if (sourceTableHandle.getSchemaTableName().equals(sourceSchemaTable)) {
```

Comment:
can there be a mismatch really? Please add a comment

Reply:
There shouldn't be a mismatch in theory: that code segment that creates the snapshot summary string is in `finishRefreshMaterializedView()`, and it uses the `sourceTableHandles` list there too, so it looks to be correct. Therefore, I'm removing this extra check from here.